### PR TITLE
#182282571: Start django app - users

### PR DESCRIPTION
**Pull request by narravabrion**
#### What does this Pull Request do?
+ start a user app

#### Description of Task to be completed?
+ start user app
+ create custom user model
+ register model to admin panel

#### How should this be manually tested?
+ unittests for models

#### Any background context you want to provide?
+ create a custom user extending the AbstractBaseUser and BaseUserManager

#### What are the relevant Pivotal tracker card?
+ [Start django app - users](https://www.pivotaltracker.com/n/projects/2570604/stories/182282571)

#### Screenshots (if appropriate)
N/A

#### Questions:
N/A

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,10 +1,33 @@
...
 click==8.1.3
+colorama==0.4.4
+distlib==0.3.4
 Django==4.0.5
 djangorestframework==3.13.1
+filelock==3.7.1
 flake8==4.0.1
+identify==2.5.1
 isort==5.10.1
-mypy==0.812
```
Comment (MwanikiN) [inline requirements.txt:7]: mypy==0.961. Latest version.

File: requirements.txt [hunk 2]
```diff
@@ -1,10 +1,33 @@
...
+platformdirs==2.5.2
 pre-commit==2.19.0
+psycopg2-binary==2.9.3
+pycodestyle==2.8.0
+pyflakes==2.4.0
+python-dotenv==0.20.0
+pytz==2022.1
+PyYAML==6.0
 six==1.16.0
-psycopg2-binary
```
Comment (MwanikiN) [inline requirements.txt:10]: psycopg2--binary   Corrected a missing (-)

File: core/settings/base.py [hunk 3]
```diff
@@ -96,3 +97,10 @@
 
 
 DEFAULT_AUTO_FIELD = "django.db.models.BigAutoField"
+
+# file settings
+MEDIA_URL = "/media/"
+MEDIA_ROOT = os.path.join(BASE_DIR, "media")
+
+# custom user
+AUTH_USER_MODEL = 'users.User'
```
Comment (mainamuriuki274) [inline core/settings/base.py]: add an emplty line after this file

File: users/admin.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+from django.contrib import admin
+from django.contrib.auth import get_user_model
+
+# Register your models here.
```
Comment (mainamuriuki274) [inline users/admin.py]: remove this line

File: users/admin.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
...
+User =  get_user_model()
+admin.site.register(User)
```
Comment (mainamuriuki274) [inline users/admin.py]: add an empty line after this line

File: users/models.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+from django.db import models
+from django.contrib.auth.models import BaseUserManager
+from django.contrib.auth.models import AbstractBaseUser, PermissionsMixin
+
+
+# Create your models here.
```
Comment (mainamuriuki274) [inline users/models.py]: remove this linne

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 1 above)
Comment (narravabrion) [reply requirements.txt:7]: Use mypy==0.950

File: core/settings/base.py (hunk 3 above)
Comment (narravabrion) [reply core/settings/base.py]: done

File: users/admin.py (hunk 4 above)
Comment (narravabrion) [reply users/admin.py]: done

File: users/admin.py (hunk 5 above)
Comment (narravabrion) [reply users/admin.py]: done

File: users/models.py (hunk 6 above)
Comment (narravabrion) [reply users/models.py]: done

File: users/models.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    objects = UserManager()
+
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["username"]
+
+    def __str__(self) -> str:
+        return self.email
+
+
+class Profile(models.Model):
```
Comment (mainamuriuki274) [inline users/models.py]: create an abstract TimeStampedModel that is inherited in all models

File: users/models.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        self, email: str, password: str, **kwargs: Any
+    ) -> Any:
+        user = self.create_user(email, password=password, **kwargs)
+        user.is_staff = True
+        user.is_superuser = True
+        user.save(using=self._db)
+        return user
+
+
+class User(AbstractBaseUser, PermissionsMixin):
```
Comment (mainamuriuki274) [inline users/models.py]: create an abstract TimeStampedModel that is inherited in all models

File: users/models.py [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+        user.save(using=self._db)
+        return user
+
+    def create_superuser(
+        self, email: str, password: str, **kwargs: Any
+    ) -> Any:
+        user = self.create_user(email, password=password, **kwargs)
+        user.is_staff = True
+        user.is_superuser = True
+        user.save(using=self._db)
```
Comment (mainamuriuki274) [inline users/models.py]: you can remove using=self._db

File: users/models.py [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+        user = self.model(email=self.normalize_email(email).lower(), **kwargs)
+        user.set_password(password)
+        user.save(using=self._db)
+        return user
+
+    def create_editor(self, email: str, password: str, **kwargs: Any) -> Any:
+        user = self.create_user(email, password=password, **kwargs)
+        user.is_editor = True
+        user.save(using=self._db)
```
Comment (mainamuriuki274) [inline users/models.py]: you can remove using=self._db

File: users/models.py [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+class UserManager(BaseUserManager):
+    def create_user(self, email: str, password: str, **kwargs: Any) -> Any:
+        if not email:
+            raise ValueError("Email is required")
+        if not password:
+            raise ValueError("Password is required")
+
+        user = self.model(email=self.normalize_email(email).lower(), **kwargs)
+        user.set_password(password)
+        user.save(using=self._db)
```
Comment (mainamuriuki274) [inline users/models.py]: you can remove using=self._db

File: .vscode/settings.json [hunk 12]
```diff
@@ -0,0 +1,3 @@
+{
+    "python.formatting.provider": "black"
```
Comment (mainamuriuki274) [inline .vscode/settings.json]: add .vscode to .gitignore

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 13]
```diff
@@ -0,0 +1,3 @@
...
+}
```
Comment (mainamuriuki274) [inline .vscode/settings.json]: You need to delete this file also

File: users/models.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+    objects = UserManager()
+
+    USERNAME_FIELD = "email"
+    REQUIRED_FIELDS = ["username"]
+
+    def __str__(self) -> str:
+        return self.email
+
+
+class Profile(TimeStampedModel, models.Model):
```
Comment (mainamuriuki274) [inline users/models.py]: if you've inherited TimeStampedModel , you do not need to inherit models.Model

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

**Review by mainamuriuki274 — APPROVED** [review]

**Pull request merged**